Mathlib/NumberTheory/Frobenius/Basic.lean

```diff
@@ -0,0 +1,315 @@
+cheaply from `A[X]` to `B[X]` given that `B` is an `A`-algebra.  -/
+noncomputable local instance : Algebra A[X] B[X] :=
+  RingHom.toAlgebra (Polynomial.mapRingHom (Algebra.toRingHom))
+
+@[simp, norm_cast]
+lemma _root_.Polynomial.coe_monomial' (n : ℕ) (a : A) :
+    ((monomial n a : A[X]) : B[X]) = monomial n (a : B) :=
+  map_monomial _
+
+lemma F.descent (h : ∀ b : B, (∀ σ : B ≃ₐ[A] B, σ • b = b) → ∃ a : A, b = a) :
+    ∃ m : A[X], (m : B[X]) = F A Q := by
+  choose f hf using h
+  classical
+  let f' : B → A := fun b ↦ if h : ∀ σ : B ≃ₐ[A] B, σ b = b then f b h else 37
+  use (∑ x ∈ (F A Q).support, (monomial x) (f' ((F A Q).coeff x)))
+  ext N
+  push_cast
+  simp_rw [finset_sum_coeff, ← lcoeff_apply, lcoeff_apply, coeff_monomial,
+    Finset.sum_ite_eq', mem_support_iff, ne_eq, ite_not, f']
+  split
+  · next h => exact h.symm
+  · next h1 =>
+    rw [dif_pos <| fun σ ↦ ?_]
+    · refine (hf ?_ ?_).symm
+    · nth_rw 2 [← F.smul_eq_self σ]
+      rfl
+
+-- This will be true in applications to number fields etc. A Galois-invariant element
+-- of `B` is in `A`.
+variable (isGalois : ∀ b : B, (∀ σ : B ≃ₐ[A] B, σ • b = b) → ∃ a : A, b = a)
+
+/-- An auxiliary polynomial, needed in the definition of Frobenius elements.
+It's the descent to `A[X]` of a certain auxiliary Galois-invariant polynomial in `B[X]`. -/
+noncomputable abbrev m := (F.descent A Q isGalois).choose
+
+lemma m_spec : ((m A Q isGalois) : B[X]) = F A Q := (F.descent A Q isGalois).choose_spec
+
+lemma m_spec' : (m A Q isGalois).map (algebraMap A B) = F A Q := by
+  rw [← m_spec A Q isGalois]
+  rfl
+
+-- Amelia's trick to insert "let P be the ideal under Q" into the typeclass system
+variable (P : Ideal A) [P.IsMaximal] [Algebra (A ⧸ P) (B ⧸ Q)] [IsScalarTower A (A⧸P) (B⧸Q)]
+variable {A} -- now can be implicit because `P`
+
+lemma m.y_mod_P_eq_zero : Polynomial.aeval (↑(y A Q) : B ⧸ Q) (m A Q isGalois) = 0 := by
+  rw [← aeval_map_algebraMap B, m_spec', algebraMap.coe_def, aeval_algebraMap_apply,
+    coe_aeval_eq_eval, F.y_eq_zero A Q, map_zero]
+
+open scoped Polynomial
+
+variable [Fintype (A⧸P)]
+
+lemma m.mod_P_y_pow_q_eq_zero' :
+    aeval ((algebraMap B (B⧸Q) (y A Q)) ^ (Fintype.card (A⧸P)))  (m A Q isGalois) = 0 := by
+  let foobar : Field (A⧸P) := ((Ideal.Quotient.maximal_ideal_iff_isField_quotient P).mp ‹_›).toField
+  rw [← aeval_map_algebraMap (A⧸P), FiniteField.aeval_pow_card, ← algebraMap.coe_def,
+    aeval_map_algebraMap, m.y_mod_P_eq_zero, zero_pow Fintype.card_ne_zero]
+
+lemma F.mod_Q_y_pow_q_eq_zero' :
+    aeval ((algebraMap B (B⧸Q) (y A Q)) ^ (Fintype.card (A⧸P))) (F A Q) = 0 := by
+  rw [← m_spec' A Q isGalois, aeval_map_algebraMap, m.mod_P_y_pow_q_eq_zero']
+
+lemma _root_.Polynomial.aeval_finset_prod.{u, v, y} {R : Type u} {S : Type v} {ι : Type y}
+    [CommSemiring R] [CommSemiring S] [Algebra R S] (s : Finset ι) (g : ι → R[X]) (x : S) :
+  aeval x (∏ i ∈ s, g i) = (∏ i ∈ s, aeval x (g i)) := eval₂_finset_prod (algebraMap R S) s g x
```

Comment:
```suggestion
    aeval x (∏ i ∈ s, g i) = (∏ i ∈ s, aeval x (g i)) := eval₂_finset_prod (algebraMap R S) s g x
```

Reply:
I have mistakenly left this lemma here -- probably it should go elsewhere.